src/charm.py

```diff
@@ -121,92 +123,107 @@ def _get_database_relation_info(self) -> dict:
-            logger.error(str(err))
-            self.unit.status = BlockedStatus(str(err))
-        else:
-            if current_config != self._config:
-                self._container.push(self._hydra_config_path, self._config, make_dirs=True)
-                logger.info("Updated hydra config, restarting the container")
-                self._container.restart()
-
-    def _update_container(self, event) -> None:
-        """Update configs, pebble layer and run database migration."""
-        self.unit.status = MaintenanceStatus("Configuring hydra container")
+            stdout, _ = process.wait_output()
+            logger.info(f"Executing automigration: {stdout}")
+        except ExecError as err:
+            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
+            self.unit.status = BlockedStatus("Database migration job failed")
 
+    def _on_hydra_pebble_ready(self, event) -> None:
+        """Event Handler for pebble ready event."""
         if not self.model.relations["pg-database"]:
             # TODO: Observe relation_joined event instead of event deferring
             event.defer()
             logger.error("Missing required relation with postgresql")
             self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
             return
 
-        if not self.database.is_database_created():
-            event.defer()
-            logger.info("Missing database details. Deferring the event.")
-            self.unit.status = WaitingStatus("Waiting for database creation")
-            return
-
         if not self._container.can_connect():
             event.defer()
             logger.info("Cannot connect to Hydra container. Deferring the event.")
             self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
             return
 
-        self._update_layer()
+        self.unit.status = MaintenanceStatus("Configuring resources")
+
+        self._container.add_layer(self._container_name, self._hydra_layer, combine=True)
+        logger.info("Pebble plan updated with new configuration, replanning")
+        try:
+            self._container.replan()
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus("Failed to replan")
+            return
+
+        if self.database.is_database_created():
+            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+            self._container.start(self._container_name)
+            self.unit.status = ActiveStatus()
+
+    def _on_database_created(self, event) -> None:
+        """Event Handler for database created event."""
+        logger.info("Retrieved database details")
 
         if not self.unit.is_leader():
             # TODO: Observe leader_elected event
             logger.info("Unit does not have leadership")
             self.unit.status = WaitingStatus("Waiting for leadership")
```

Comment:
If a unit is not the leader, it does not mean that there is no leader. You should not change the status of the unit here.

Reply:
I'm not sure I understand. The waiting status is set only to non-leader unit, not the entire charm. Is that wrong?

Reply:
The message is wrong. The unit is not waiting for leadership, it is waiting for the leader to run the migration. I am fine with changing the status to waiting, but we should have a better message for it.

Reply:
Ok, message changed. Thanks for explanation